src/app/app.cpp

```diff
@@ -963,11 +977,38 @@ namespace app
     void fill_header(
       etcdserverpb::ResponseHeader& header, const ccf::TxID& tx_id)
     {
+      header.set_cluster_id(cluster_id);
       header.set_member_id(member_id());
       header.set_raft_term(tx_id.view);
       header.set_revision(tx_id.seqno);
     }
 
+    void populate_cluster_id(kv::ReadOnlyTx& tx)
+    {
+      cluster_id = get_cluster_id(tx);
+    }
+
+    int64_t get_cluster_id(kv::ReadOnlyTx& tx)
+    {
+      auto ccf_governance_map =
+        tx.template ro<ccf::Service>(ccf::Tables::SERVICE);
+      auto service_info = ccf_governance_map->get();
+
+      if (!service_info.has_value())
+      {
+        // shouldn't but just in case
+        CCF_APP_FAIL("Failed to get id for cluster");
+        return 0;
+      }
+
+      auto cert = service_info.value().cert;
+      auto sha = crypto::sha256(cert.raw());
```

Comment:
This should be derived from the certificate public key instead, as members[ can renew the service certificate](https://microsoft.github.io/CCF/main/operations/certificates.html#service-certificate), which would change the cluster id. The public key should be extractable via `crypto::make_verifier(cert)->public_key_der()`.